repo2docker/app.py

```diff
@@ -560,10 +560,103 @@ def find_image(self):
                         return True
         return False
 
+    def get_r2d_version(self, checkout_path, default_buildpack):
+        """
+        Read the repo2docker.version if present, otherwise return the 
+        current version. Sanity check that an image with the specified 
+        version exists.
+        """
+
+        # Use the default buildpack, since there's a convenience method to read
+        # the binder_path
+        r2d_version_path = default_buildpack.binder_path("repo2docker.version")
+
+        # Read the version file, otherwise default to the current version
+        if os.path.exists(r2d_version_path):
+
+            # This assumes that the file contains the version on a single line
+            with open(r2d_version_path, "r") as f:
+                version = f.readline().strip()
+
+            # Preliminary sanity check that an image with the specified tag version
+            url = "https://hub.docker.com/v2/repositories/jupyter/repo2docker/tags/{}".format(version)
+
+            try:
+                urllib.request.urlopen(url)
+            except urllib.error.HTTPError as e:
+                self.log.error("\nUnable to find requested version %s (%s)\n",
+                               version, e.code, extra=dict(phase="failure"))
+                raise
+        else:
+            version = self.version
+
+        self.log.info("\nRepo specifies version %s, current is %s\n",
+                      version, self.version)
+        return version
+
+    def run_r2d_version(self, version):
+        """
+        Run a previous version of repo2docker via Docker
+        """
+
+        # Stored unparsed argument list is modified below
+        args = self.argv
+
+        # If the user didn't specify an image name, then the generated name will
+        # differ between the current repo2docker session and the call via
+        # container. In this case, inject the name generated in this session as 
+        # an argument. 
+        if '--image-name' not in args:
+            args[0:0] = ['--image-name', self.output_image_spec]
+
+        # Map of volumes to be mounted into the repo2docker container. The
+        # docker socket will always be mounted.
+        volumes = {
+            "/var/run/docker.sock": {"bind": "/var/run/docker.sock", "mode": "rw"}
+        }
+
+        if os.path.isdir(self.repo):
+            # If the specified repo is a local path, we'll use a fixed path
+            # in the container.
+            args[-1] = '/repo2docker'
+
+            # Ensure the path is absolute 
+            if os.path.isabs(self.repo):
+                source_path = self.repo
+            else:
+                source_path = os.path.abspath(self.repo)
```

Comment:
I'd just always call `source_path = os.path.abspath(self.repo)` since for absolute path it's a noop.